clang/lib/Driver/ToolChains/SYCL.cpp

```diff
@@ -417,10 +416,96 @@ SYCLToolChain::SYCLToolChain(const Driver &D, const llvm::Triple &Triple,
+      {"libc", true}, {"libm-fp32", true}, {"libm-fp64", false}};
+  if (Arg *A = Args.getLastArg(options::OPT_fsycl_device_lib_EQ,
+                               options::OPT_fno_sycl_device_lib_EQ)) {
+    if (A->getValues().size() == 0)
+      getDriver().Diag(diag::warn_drv_empty_joined_argument)
+          << A->getAsString(Args);
+    else {
+      if (A->getOption().matches(options::OPT_fno_sycl_device_lib_EQ))
+        NoDeviceLibs = true;
+
+      for (StringRef Val : A->getValues()) {
+        if (Val == "all") {
+          for (auto &K : devicelib_link_info.keys())
+            devicelib_link_info[K] = true && !NoDeviceLibs;
+          break;
+        }
+        auto LinkInfoIter = devicelib_link_info.find(Val);
+        if (LinkInfoIter == devicelib_link_info.end()) {
+          getDriver().Diag(diag::err_drv_unsupported_option_argument)
+              << A->getOption().getName() << Val;
+        }
+        devicelib_link_info[Val] = true && !NoDeviceLibs;
+      }
+    }
+  }
+  bool isSpirvAOT = true;
+  SmallString<128> LibLoc(getDriver().Dir);
+  llvm::sys::path::append(LibLoc, "/../lib");
+  SmallVector<DeviceLibOptInfo, 5> sycl_device_wrapper_libs = {
+      {"libsycl-crt", "libc"},
+      {"libsycl-complex", "libm-fp32"},
+      {"libsycl-complex-fp64", "libm-fp64"},
+      {"libsycl-cmath", "libm-fp32"},
+      {"libsycl-cmath-fp64", "libm-fp64"}};
+  // For AOT compilation, we need to link sycl_device_fallback_libs as
+  // default too.
+  SmallVector<DeviceLibOptInfo, 5> sycl_device_fallback_libs = {
+      {"libsycl-fallback-cassert", "libc"},
+      {"libsycl-fallback-complex", "libm-fp32"},
+      {"libsycl-fallback-complex-fp64", "libm-fp64"},
+      {"libsycl-fallback-cmath", "libm-fp32"},
+      {"libsycl-fallback-cmath-fp64", "libm-fp64"}};
+
+  auto addInputs = [&](SYCLDeviceLibType t, StringRef SYCLArchName) {
+    auto sycl_libs = (t == sycl_devicelib_wrapper) ? sycl_device_wrapper_libs
+                                                   : sycl_device_fallback_libs;
+    for (const DeviceLibOptInfo &Lib : sycl_libs) {
+      if (!devicelib_link_info[Lib.devicelib_option])
+        continue;
+      SmallString<128> LibName(LibLoc);
+      llvm::sys::path::append(LibName, Lib.devicelib_name);
+      LibName.append("-");
+      LibName.append(SYCLArchName);
+      llvm::sys::path::replace_extension(LibName, ".bc");
+      if (llvm::sys::fs::exists(LibName)) {
+        CC1Args.push_back("-mlink-builtin-bitcode");
+        CC1Args.push_back(Args.MakeArgString(LibName.c_str()));
+      }
+    }
+  };
+  addInputs(sycl_devicelib_wrapper, getTriple().getArchName());
+  if (isSpirvAOT)
+    addInputs(sycl_devicelib_fallback, getTriple().getArchName());
+}
+
 void SYCLToolChain::addClangTargetOptions(
     const llvm::opt::ArgList &DriverArgs,
     llvm::opt::ArgStringList &CC1Args,
     Action::OffloadKind DeviceOffloadingKind) const {
+  if (DeviceOffloadingKind == Action::OFK_SYCL && doLLVMBCEmit(CC1Args))
```

Comment:
Hi, @bader 
If we don't do this check, "-mlink-builtin-bitcode ..." will be added when we generate  sycl helper header file. We only need to link those device libraries when generating device code.

Reply:
We need to link them always. Integration header is generated from Sema, so adding `-m*` options should have no impact.

Reply:
Hi, @bader 
Since the "-mlink" options have no impact to integration header generation, why don't remove the unused options to make the commands more clean?
Thanks very much.

Reply:
This is not the right place to make this decision. We must always link devicelib unless user explicitly disable the linking. We should have some "optimizing" logic here as it's a error prone. I can image a few cases already where this can lead to bugs.